Foundation/Decimal.swift

```diff
@@ -1691,10 +1639,71 @@ public func NSDecimalMultiplyByPowerOf10(_ result: UnsafeMutablePointer<Decimal>
 }
 
 public func NSDecimalString(_ dcm: UnsafePointer<Decimal>, _ locale: Any?) -> String {
-    guard locale == nil else {
-        fatalError("Locale not supported: \(locale!)")
+
+    var copy = dcm.pointee
+    if copy.isNaN {
+        return "NaN"
+    }
+    if copy._length == 0 {
+        return "0"
+    }
+
+    let ZERO: CChar = 0x30 // ASCII '0' == 0x30
+    let DECIMALPOINT: CChar = 0x2e // ASCII '.' == 0x2e
+    let MINUS: CChar = 0x2d // ASCII '-' == 0x2d
+    let bufferSize = 200 // Max value: 39 + 128 + sign + decimal point
+    var buffer = Array<CChar>(repeating: 0, count: bufferSize)
+
+    var i = bufferSize - 1
+    while copy._exponent > 0 {
+        i -= 1
+        buffer[i] = ZERO
+        copy._exponent -= 1
+    }
+
+    if copy._exponent == 0 {
+        copy._exponent = 1
     }
-    return dcm.pointee.description
+
+    while copy._length != 0 {
+        var remainder: UInt16 = 0
+        if copy._exponent == 0 {
+            i -= 1
+            buffer[i] = DECIMALPOINT
+        }
+        copy._exponent += 1
+        (remainder, _) = divideByShort(&copy, 10)
+        i -= 1
+        buffer[i] = Int8(remainder) + ZERO
+    }
+    if copy._exponent <= 0 {
+        while copy._exponent != 0 {
+            i -= 1
+            buffer[i] = ZERO
+            copy._exponent += 1
+        }
+        i -= 1
+        buffer[i] = DECIMALPOINT
+        i -= 1
+        buffer[i] = ZERO
+    }
+    if copy._isNegative != 0 {
+        i -= 1
+        buffer[i] = MINUS
+    }
+    let result = String(cString: Array(buffer.suffix(from: i)))
+
+    var decimalSeparator: String?
+    if let locale = locale as? Locale {
+        decimalSeparator = locale.decimalSeparator
+    } else if let dictionary = locale as? Dictionary<NSLocale.Key, String> {
```

Comment:
Should this be `[NSLocale.Key: Any]`?

Note that unlike Darwin, `NSLocale.Key` will not automatically bridge from `String` or `NSString`. Should we have a branch for that too?

Reply:
I added a branch for `[String: String]`. 

For the `[NSLocale.Key: Any]` case wouldn't that then mean `as!` casting the value to `String`? It that what you want it changed to?

Reply:
No, the current patch is good as-is.

Reply:
(After adding the branch.)